examples/fabric/dcgan/train_fabric.py

```diff
@@ -0,0 +1,268 @@
+# Number of channels in the training images
+nc = 3
+# Size of z latent vector (i.e. size of generator input)
+nz = 100
+# Size of feature maps in generator
+ngf = 64
+# Size of feature maps in discriminator
+ndf = 64
+# Number of training epochs
+num_epochs = 5
+# Learning rate for optimizers
+lr = 0.0002
+# Beta1 hyperparameter for Adam optimizers
+beta1 = 0.5
+# Number of GPUs to use
+num_gpus = 1
+
+
+def main():
+    # Set random seed for reproducibility
+    seed_everything(999)
+
+    fabric = Fabric(accelerator="auto", devices=1)
+    fabric.launch()
+
+    dataset = CelebA(
+        root=dataroot,
+        split="all",
+        download=True,
+        transform=transforms.Compose(
+            [
+                transforms.Resize(image_size),
+                transforms.CenterCrop(image_size),
+                transforms.ToTensor(),
+                transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),
+            ]
+        ),
+    )
+
+    # Create the dataloader
+    dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, shuffle=True, num_workers=workers)
+
+    output_dir = Path("outputs-fabric", time.strftime("%Y%m%d-%H%M%S"))
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+    # Plot some training images
+    real_batch = next(iter(dataloader))
+    torchvision.utils.save_image(
+        real_batch[0][:64],
+        output_dir / "sample-data.png",
+        padding=2,
+        normalize=True,
+    )
+
+    # Create the generator
+    generator = Generator()
+
+    # Apply the weights_init function to randomly initialize all weights
+    generator.apply(weights_init)
+
+    # Create the Discriminator
+    discriminator = Discriminator()
+
+    # Apply the weights_init function to randomly initialize all weights
+    discriminator.apply(weights_init)
+
+    # Initialize BCELoss function
+    criterion = nn.BCELoss()
+
+    # Create batch of latent vectors that we will use to visualize
+    #  the progression of the generator
+    fixed_noise = torch.randn(64, nz, 1, 1, device=fabric.device)
+
+    # Establish convention for real and fake labels during training
+    real_label = 1.0
+    fake_label = 0.0
+
+    # Set up Adam optimizers for both G and D
+    optimizer_d = optim.Adam(discriminator.parameters(), lr=lr, betas=(beta1, 0.999))
+    optimizer_g = optim.Adam(generator.parameters(), lr=lr, betas=(beta1, 0.999))
+
+    discriminator, optimizer_d = fabric.setup(discriminator, optimizer_d)
+    generator, optimizer_g = fabric.setup(generator, optimizer_g)
+    dataloader = fabric.setup_dataloaders(dataloader)
+
+    # Lists to keep track of progress
+    losses_g = []
+    losses_d = []
+    iteration = 0
+
+    # Training loop
+    for epoch in range(num_epochs):
+        for i, data in enumerate(dataloader, 0):
+            # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
+            # (a) Train with all-real batch
+            discriminator.zero_grad()
+            real_cpu = data[0]
```

Comment:
why CPU? it is confusing name

Reply:
```suggestion
            real_spl = data[0]
```

Reply:
because in train_torch.py, they moved it to the device later on.